#! /bin/sh /usr/share/dpatch/dpatch-run
## 168_rexml_dos.dpatch by Lucas Nussbaum <lucas@lucas-nussbaum.net>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Backported upstream r19033 (from trunk).
## DP: Fixes CVE-2008-3790 (REXML expansion DOS). Closes: #496808

@DPATCH@
diff -urNad ruby1.8-1.8.7.72~/lib/rexml/document.rb ruby1.8-1.8.7.72/lib/rexml/document.rb
--- ruby1.8-1.8.7.72~/lib/rexml/document.rb	2008-06-06 10:05:24.000000000 +0200
+++ ruby1.8-1.8.7.72/lib/rexml/document.rb	2008-09-10 09:39:07.000000000 +0200
@@ -32,6 +32,7 @@
 	  # @param context if supplied, contains the context of the document;
 	  # this should be a Hash.
 		def initialize( source = nil, context = {} )
+			@entity_expansion_count = 0
 			super()
 			@context = context
 			return if source.nil?
@@ -200,6 +201,27 @@
 			Parsers::StreamParser.new( source, listener ).parse
 		end
 
+		@@entity_expansion_limit = 10_000
+
+		# Set the entity expansion limit. By defualt the limit is set to 10000.
+		def Document::entity_expansion_limit=( val )
+		  @@entity_expansion_limit = val
+		end
+
+		# Get the entity expansion limit. By defualt the limit is set to 10000.
+		def Document::entity_expansion_limit
+		  return @@entity_expansion_limit
+		end
+
+		attr_reader :entity_expansion_count
+
+		def record_entity_expansion
+		  @entity_expansion_count += 1
+		  if @entity_expansion_count > @@entity_expansion_limit
+		    raise "number of entity expansions exceeded, processing aborted."
+		  end
+		end
+
 		private
 		def build( source )
       Parsers::TreeParser.new( source, self ).parse
diff -urNad ruby1.8-1.8.7.72~/lib/rexml/entity.rb ruby1.8-1.8.7.72/lib/rexml/entity.rb
--- ruby1.8-1.8.7.72~/lib/rexml/entity.rb	2008-04-18 09:22:13.000000000 +0200
+++ ruby1.8-1.8.7.72/lib/rexml/entity.rb	2008-09-10 09:39:07.000000000 +0200
@@ -73,6 +73,7 @@
 		# all entities -- both %ent; and &ent; entities.  This differs from
 		# +value()+ in that +value+ only replaces %ent; entities.
 		def unnormalized
+			document.record_entity_expansion unless document.nil?
 			v = value()
 			return nil if v.nil?
 			@unnormalized = Text::unnormalize(v, parent)
